src/Features/Core/Portable/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -0,0 +1,243 @@
+            if (!options.TryGetEditorConfigOption(CodeStyleOptions.FileHeaderTemplate, out string fileHeaderTemplate)
+                || string.IsNullOrEmpty(fileHeaderTemplate))
+            {
+                // Avoid making changes if we fail to read the expected header
+                return root;
+            }
+
+            var expectedFileHeader = fileHeaderTemplate.Replace("{fileName}", Path.GetFileName(document.FilePath));
+
+            var fileHeader = FileHeaderHelper.ParseFileHeader(root);
+            SyntaxNode newSyntaxRoot;
+            if (fileHeader.IsMissing)
+            {
+                newSyntaxRoot = AddHeader(document, root, expectedFileHeader);
+            }
+            else
+            {
+                newSyntaxRoot = ReplaceHeader(document, root, expectedFileHeader);
+            }
+
+            return newSyntaxRoot;
+        }
+
+        private SyntaxNode ReplaceHeader(Document document, SyntaxNode root, string expectedFileHeader)
+        {
+            // Skip single line comments, whitespace, and end of line trivia until a blank line is encountered.
+            var trivia = root.GetLeadingTrivia();
+            var onBlankLine = false;
+            var inCopyright = true;
+            var removalList = new List<int>();
+            var leadingSpaces = string.Empty;
+            var possibleLeadingSpaces = string.Empty;
+
+            // Need to do this with index so we get the line endings correct.
+            for (var i = 0; i < trivia.Count; i++)
+            {
+                var triviaLine = trivia[i];
+                if (triviaLine.RawKind == SyntaxKinds.SingleLineCommentTrivia)
+                {
+                    if (possibleLeadingSpaces != string.Empty)
+                    {
+                        leadingSpaces = possibleLeadingSpaces;
+                    }
+
+                    removalList.Add(i);
+                    onBlankLine = false;
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.WhitespaceTrivia)
+                {
+                    if (leadingSpaces == string.Empty)
+                    {
+                        possibleLeadingSpaces = triviaLine.ToFullString();
+                    }
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+                }
+                else if (triviaLine.RawKind == SyntaxKinds.EndOfLineTrivia)
+                {
+                    possibleLeadingSpaces = string.Empty;
+
+                    if (inCopyright)
+                    {
+                        removalList.Add(i);
+                    }
+
+                    if (onBlankLine)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        onBlankLine = true;
+                    }
```

Comment:
fwiw... i prefer my way of matching banners.  effectively to use regex-esque patterns over syntax-trivia-lits.  THe reason being is that it can effectively express that we need more complicated state here for.  i.e. if i'm reading this correctly, this is just trying to say "stop once i get `<new-line>space*<new-line>`.  In that sort of form it's easy to read.  In this loop, i find it much harder to figure out the interplay of all the state variables.

Reply:
I've added a bunch of comments. There may be room for future improvement but I'm happy with the current behavior and test coverage as a starting point.

Reply:
yup.  im fine withthis now.  it's much easier to understand.  thanks!